### PR TITLE
Add variables argument to the DatacubeExtension `apply` method

### LINKED ISSUE #781: DatacubeExtension.apply is unaware of variables
The recent addition (#699) of the variable setter did not add an accompanying change to the `DatacubeExtension.apply` method. As a result, you cannot set dimensions and variables in one swoop with the apply method. The [documentation implies](https://pystac.readthedocs.io/en/latest/concepts.html#accessing-extension-functionality) that `apply` should allow you to pass in extension properties.

Additionally, there is no test for the apply method, so adding one seems appropriate now and will help with #324.

### PR DESCRIPTION
**Pull request by itcarroll**
**Related Issue(s):**

- Fixes #781
- Helps with #324

**Description:**

An optional argument is added to the `DatacubeExtension.apply` method called `variables` and handed off to the variable setter. The argument is optional because Variable Objects are optional in the datacube extension spec.

A new `test_apply_variables` method for the `DatacubeTest` class does a round-trip check on a variable present in the example item using `apply`.

**PR Checklist:**

- [x] Code is formatted (run `pre-commit run --all-files`)
- [x] Tests pass (run `scripts/test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

**Comment by duckontheweb** [comment]
> I don't have a lot of experience with type hints, so I'm not sure what to do about the following test failure:
> 
> ```
> mypy.....................................................................Failed
> - hook id: mypy
> - exit code: 1
> 
> tests/extensions/test_datacube.py:94: error: Item "None" of "Optional[Dict[str, Variable]]" has no attribute "popitem"  [union-attr]
> tests/extensions/test_datacube.py:98: error: Value of type "None" is not indexable  [index]
> ```
> 
> Is there a convention for checking a `@property` for None, or using try/except, that I should employ here?

We have mostly been handling this within the tests by using something like the following:

```python
cube = DatacubeExtension.ext(item)
cube_variables = cube.variables
assert cube_variables is not None
key, value = cube_variables.popitem()
```

It's a bit verbose, but `mypy` picks up on the `assert` statement and knows that it can't be `None` after that point.

**Comment by itcarroll** [comment]
I'll include a commit updating the changelog when/if workflow approval is granted, which I presume will do the coverage check.

**Review by duckontheweb — APPROVED** [review]
Thanks for catching and fixing this @itcarroll! I'll merge once the CHANGELOG has been updated.

**Comment by itcarroll** [comment]
Added to the CHANGELOG. It looks like the coverage workflow still requires approval to run.

**Comment by itcarroll** [comment]
Coverage at [8964f5c](https://codecov.io/gh/stac-utils/pystac/commit/8964f5cc0a075a9ada53f79ca14b02029d9e89bb) at 94.58 equivalent to main at [07816f5e](https://app.codecov.io/gh/stac-utils/pystac/commit/07816f5e85b67750ad8dc2d84c446c5ed3233602). Not sure why Codecov didn't report back here.

**Comment by duckontheweb** [comment]
> Coverage at [8964f5c](https://codecov.io/gh/stac-utils/pystac/commit/8964f5cc0a075a9ada53f79ca14b02029d9e89bb) at 94.58 equivalent to main at [07816f5e](https://app.codecov.io/gh/stac-utils/pystac/commit/07816f5e85b67750ad8dc2d84c446c5ed3233602). Not sure why Codecov didn't report back here.

Yeah, I've noticed really long lag times in Codecov posting back to the PR and sometimes it seems to take a few CI runs before it triggers. Looks like things are good on that front, thanks for the work on this!

**Pull request merged**
